neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -246,17 +257,93 @@ func QmpQueryMemoryDevices(ip string, port int32) ([]QmpMemoryDevice, error) {
 	return result.Return, nil
 }
 
+func QmpMatchName(reg *regexp.Regexp, name string) (int, bool) {
+	// extracts the id from the name
+	matches := reg.FindStringSubmatch(name)
+	if len(matches) < 2 {
+		return 0, false
+	}
+	id, err := strconv.Atoi(matches[1])
+	if err != nil {
+		return 0, false
+	}
+	return id, true
+}
+
+var memBackendRegexp = regexp.MustCompile(`memslot(\d+)`)
+
+func QmpMatchMemBackend(name string) (int, error) {
+	// "/objects/memslot3" -> 3
+	name = strings.ReplaceAll(name, "/objects/", "")
+	idx, ok := QmpMatchName(memBackendRegexp, name)
+	if !ok {
+		return 0, fmt.Errorf("failed to parse memory device id: %s", name)
+	}
+	return idx, nil
+}
+
+func QmpQueryMemoryBackendIds(mon *qmp.SocketMonitor) (map[int]struct{}, error) {
+	cmd := []byte(`{"execute": "qom-list", "arguments": {"path": "/objects"}}`)
+	raw, err := mon.Run(cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	var result QmpObjects
+	json.Unmarshal(raw, &result)
+	backends := map[int]struct{}{}
+	for _, o := range result.Return {
+		idx, err := QmpMatchMemBackend(o.Name)
+		if err == nil {
+			backends[idx] = struct{}{}
+		}
+	}
+	return backends, nil
+}
+
+func QmpAddMemoryBackend(mon *qmp.SocketMonitor, idx int, size int64) error {
+	cmd := []byte(fmt.Sprintf(`{"execute": "object-add",
+								"arguments": {"id": "memslot%d",
+											  "size": %d,
+											  "qom-type": "memory-backend-ram"}}`, idx, size))
+	_, err := mon.Run(cmd)
+	return err
+}
+
+func QmpDelMemoryBackend(mon *qmp.SocketMonitor, idx int) error {
+	cmd := []byte(fmt.Sprintf(`{"execute": "object-del",
+								"arguments": {"id": "memslot%d"}}`, idx))
+	_, err := mon.Run(cmd)
+	return err
+}
+
+func QmpAddMemoryDevice(mon *qmp.SocketMonitor, idx int) error {
+	cmd := []byte(fmt.Sprintf(`{"execute": "device_add",
+								"arguments": {"id": "dimm%d",
+											  "driver": "pc-dimm",
+											  "memdev": "memslot%d"}}`, idx, idx))
+	_, err := mon.Run(cmd)
+	return err
+}
+
+func QmpDelMemoryDevice(mon *qmp.SocketMonitor, idx int) error {
+	cmd := []byte(fmt.Sprintf(`{"execute": "device_del",
+								"arguments": {"id": "dimm%d"}}`, idx))
+	_, err := mon.Run(cmd)
+	return err
+}
+
 func QmpPlugMemory(virtualmachine *vmv1.VirtualMachine) error {
 	// slots - number of pluggable memory slots (Max - Min)
-	slots := *virtualmachine.Spec.Guest.MemorySlots.Max - *virtualmachine.Spec.Guest.MemorySlots.Min
+	slots := int(*virtualmachine.Spec.Guest.MemorySlots.Max - *virtualmachine.Spec.Guest.MemorySlots.Min)
```

Comment:
this commit (ead54af9ec0010a6c8cc213b78020b46551a191b) already has a lot of changes — let's either (a) keep these variables as `int32` or (b) make the switch to `int` a separate commit.